### DISCUSSION TITLE
How to mimic useState behavior in Zustand?

### DISCUSSION THREAD

**Discussion by sammyhajomar** [pmndrs/zustand#2264] [Q&A]
Hello,

I am new to using Zustand. I have built out my store and my question is how does one mimic the function of useState from React specifically the setter?. I have following:

    tableContent: [],
    setTableContent: (tableContent) => set((state) => ({ tableContent })),

So far this only allows to replace the previous value of tableContent with a completely new one which is good for one scenario, but there is another scenario where I also want it to be able to leverage the old value when updating tableContent. I am open to having two different set functions for that if needed. Here are different scenarios of using the setRows (setTableContent):

```
const updateDays = () => {
    setRows(prevRows => calcPageLen(prevRows));
  };

  const editRow = (index: number, updatedRow: TableContentItemShape) => {
    setRows(prevRows => {
      const newRows = prevRows.map((row, i) => i === index ? updatedRow : row);
      return calcPageLen(newRows);
    });
  };

  const updateTable = (newRow: TableContentItemShape, updatedOrder: string[]) => {
    setRows(prevRows => {
      // Find the index of the new row's id in the updatedOrder array
      const rowIndex = updatedOrder.indexOf(newRow.id);
      // Insert the new row at the correct position
      const newRows = [...prevRows];
      newRows.splice(rowIndex, 0, newRow);

      return calcPageLen(newRows);
    });
  };
```

**Comment by dai-shi** [comment]
Maybe you want something like this?
```js
  tableContent: [],
  setTableContent: (fn) => set((state) => ({ tableContent: fn(state.tableContent) })),
```
